### PR TITLE
[COR-106] Add opentelmentry tracing system

### PR DESCRIPTION
**Pull request by rukolahasser**
1. Add `opentelmentry` tracing for messaging systems
2. Some metrics need to be added in another repo
3. It requires `go 1.18+` so some checks won't pass

### CODE REVIEW

**Review by cnkarz — APPROVED** [review]

**Comment by cnkarz** [comment]
some checks seem to be failing

**Comment by rukolahasser** [comment]
> some checks seem to be failing

Open telemetry requires go1.18+

File: examples/compound/go.mod [hunk 1]
```diff
@@ -7,7 +7,7 @@ replace github.com/nakji-network/connector/chain/ethereum => ../../chain/ethereu
 replace github.com/nakji-network/connector => ../../
 
 require (
-	github.com/ethereum/go-ethereum v1.10.25
+	github.com/ethereum/go-ethereum v1.10.15
```
Comment (azhang) [inline examples/compound/go.mod:10]: ?

File: connector.go [hunk 2]
```diff
@@ -89,6 +91,12 @@ func NewConnector(options ...Option) (*Connector, error) {
 	go http.ListenAndServe("0.0.0.0:8080", c.Health)
 	log.Info().Str("addr", "0.0.0.0:8080").Msg("healthcheck listening on /live and /ready")
 
+	//	Initialize trace provider
+	_, err = monitor.InitTracerProvider(context.TODO(), conf.GetString("trace.host.grpc"), c.id(), c.manifest.Version.String(), string(c.env), conf.GetFloat64("trace.sample.ratio"))
```
Comment (azhang) [inline connector.go:95]: do we want monitoring for all connectors in all environments? Do we want to allow devs to disable monitoring if they're just running it locally in dev environment?

**Review by azhang — APPROVED** [review]
A couple minor questions but overall approved

**Comment by rm289** [comment]
We do not want tracing on externally written connectors as that would couple the external devs too closely. They have no way of testing opentelemetry in our environment and I prefer we keep open telemetry use closed and not something we share.Reminder all , for 0.2.0 we are only seeking latency measurements. We are not instrumenting all code for 0.2.0. We really only need one busy connector set up to measure latency for 0.2.0On Oct 24, 2022, at 07:59, Allie ***@***.***> wrote:﻿
@azhang approved this pull request.

A couple minor questions but overall approved

In examples/compound/go.mod:
> @@ -7,7 +7,7 @@ replace github.com/nakji-network/connector/chain/ethereum => ../../chain/ethereu
 replace github.com/nakji-network/connector => ../../
 
 require (
-	github.com/ethereum/go-ethereum v1.10.25
+	github.com/ethereum/go-ethereum v1.10.15

?



In connector.go:
> @@ -89,6 +91,12 @@ func NewConnector(options ...Option) (*Connector, error) {
 	go http.ListenAndServe("0.0.0.0:8080", c.Health)
 	log.Info().Str("addr", "0.0.0.0:8080").Msg("healthcheck listening on /live and /ready")
 
+	//	Initialize trace provider
+	_, err = monitor.InitTracerProvider(context.TODO(), conf.GetString("trace.host.grpc"), c.id(), c.manifest.Version.String(), string(c.env), conf.GetFloat64("trace.sample.ratio"))

do we want monitoring for all connectors in all environments? Do we want to allow devs to disable monitoring if they're just running it locally in dev environment?

File: examples/compound/go.mod (hunk 1 above)
Comment (rukolahasser) [reply examples/compound/go.mod:10]: Be consistent with the version in `connector` lib

File: connector.go (hunk 2 above)
Comment (rukolahasser) [reply connector.go:95]: It won't crash when running locally without `opentelemetry`.

**Comment by rukolahasser** [comment]
> Maybe we should update the tests to 1.18 and 1.19 instead of 1.17

Yes. We can do that.

**Pull request merged**